gutenberg/acquire/metadata.py

```diff
@@ -162,55 +138,104 @@ def _download_metadata_archive(self):
+            for item in metadata_archive:
+                if pg_rdf_regex.search(item.name):
+                    with disable_logging():
+                        extracted = metadata_archive.extractfile(item)
+                        graph = Graph().parse(extracted)
+                    for fact in graph:
+                        if cls._metadata_is_invalid(fact):
+                            logging.info('skipping invalid triple %s', fact)
+                        else:
+                            yield fact
+
+
+class SleepycatMetadataCache(MetadataCache):
+    """Default cache manager implementation, based on Sleepycat/Berkeley DB.
+    Sleepycat is natively supported by RDFlib so this cache is reasonably fast.
 
     """
-    return any(isinstance(token, URIRef) and ' ' in token for token in fact)
+    def __init__(self, cache_location):
+        self._check_can_be_instantiated()
+        cache_uri = cache_location
+        store = 'Sleepycat'
+        MetadataCache.__init__(self, store, cache_uri)
 
+    def _populate_setup(self):
+        makedirs(self.cache_uri)
+
+    @classmethod
+    def _check_can_be_instantiated(cls):
+        try:
+            from bsddb import db
+        except ImportError:
+            try:
+                from bsddb3 import db
+            except ImportError:
+                db = None
+        if db is None:
+            raise InvalidCache('no install of bsddb/bsddb3 found')
+        del db
 
-def _iter_metadata_triples(metadata_archive_path):
-    """Yields all meta-data of Project Gutenberg texts contained in the catalog
-    dump.
+
+class SqliteMetadataCache(MetadataCache):
+    """Cache manager based on SQLite and the RDFlib plugin for SQLAlchemy.
+    Quite slow.
 
     """
-    pg_rdf_regex = re.compile(r'pg\d+.rdf$')
-    with contextlib.closing(tarfile.open(metadata_archive_path)) \
-            as metadata_archive:
-        for item in metadata_archive:
-            if re.search(pg_rdf_regex, item.name):
-                with disable_logging():
-                    graph = Graph().parse(metadata_archive.extractfile(item))
-                for fact in graph:
-                    if _metadata_is_invalid(fact):
-                        logging.info('skipping invalid triple %s', fact)
-                    else:
-                        yield fact
+    _CACHE_URI_PREFIX = 'sqlite:///'
+
+    def __init__(self, cache_location):
+        cache_uri = self._CACHE_URI_PREFIX + cache_location
+        store = plugin.get('SQLAlchemy', Store)(identifier=_DB_IDENTIFIER)
+        MetadataCache.__init__(self, store, cache_uri)
 
+    @property
+    def _local_storage_path(self):
+        return self.cache_uri[len(self._CACHE_URI_PREFIX):]
 
-_METADATA_CACHE_MANAGER = MetadataCacheManager(
-    store='Sleepycat',
-    cache_uri=local_path(os.path.join('metadata', 'metadata.db')))
 
+_METADATA_CACHE = None
 
-def set_metadata_cache_manager(cache_manager):
```

Comment:
The set_metadata_cache() function is *not* just for testing, but rather an integral part of my branch. Without a way to explicitly set a cache, you can't control where the cache is stored on the filesystem, or in the new code, control which of the two backend types you want to use. This level of control is vital when using this as part of a larger service.

For an example of what I'm trying to achieve with this module, see:
https://github.com/cpeel/openlibrary/blob/095df4b3be522eef218813a0106ee2ce7934f120/openlibrary/plugins/gutenberg/pgebook.py
and
https://github.com/cpeel/openlibrary/blob/73533a3e011074e2ffbfe416f241a6aa14fb5b6e/scripts/dev-instance/populate_gutenberg_cache.py